### PR TITLE
Cache blocking Update

### PR DESCRIPTION
**Pull request by sheino**
Allows user to specify which dimensions he does not want to block by setting block sizes to `None`.
Fixed "#pragma openmn for" bracket bug.
Fixes the cache blocking error that caused segmentation faults.
Fixes remainder loop logic. Now there will be separate remainder loop for each block. To make sure we iterate through each element.
Rewriten test_cache_blocking.py tests to cover full range of models.

### CODE REVIEW

**Comment by sheino** [comment]
@navjotk @mlange05  Inviting comments. Happy with implementation?

**Comment by mlange05** [comment]
Can you please expand on why we want this and why we need an env variable for this? Following the recent discussions about configurability and getting expected behaviour, you also really need to document what this env variables does, either in the README, the coming documentation pages, or ideally both. Also, why are we even using an env variable here if cache blocking itself is configured only through the Python API?

**Comment by sheino** [comment]
Because we do not want to be blocking on a inner most dimension by default.
No particular reason for env_var (which I have documented in README). I just thought that discussion between env_vars, config files and flags is still going on, thus I put is there temporarely.
Should I change it into flag?

**Comment by mlange05** [comment]
For now I'd stick with a flag yes, to be consistent. We want this for auto-tuning right, so setting it from Python should not be a problem?

As for the README, if we document this one particular feature, we need to document loop blocking in general first.

**Comment by sheino** [comment]
We want this both for auto_tuning and cache blocking in general. 
Will set it as a flag and will do documentation as well.

**Comment by mlange05** [comment]
This branch is missing a test for the new feature.

**Comment by sheino** [comment]
Any idea how to tests this apart from parsing generated cpp file an checking line by line whether blocks are correct?

**Comment by mlange05** [comment]
At least test for correctness if the feature is enabled, as the previous cache blocking tests do.

**Comment by sheino** [comment]
Added test and merged with master

File: README.md [hunk 1]
```diff
@@ -55,4 +55,21 @@ op = Operator(..., compiler=IntelCompiler)
...
+set `cache_blocking` flag to `True` in `Operator`. Furthermore you can
+specify the block sizes using `block_size` parameter. It can be a single
+number which will be used for all dimensions or a list explicitly stating
+block sizes for each dim(x,y,z).
+
+Note, by default inner most dimension is not blocked, if you want to
+disable this set `cb_inner_dim` flag to `True`
+Example usage:
+```
+op = Operator(..., cache_blocking=True, block_size=[5, 10, 5], cb_inner_dim=False)
```
Comment (navjotk) [inline README.md]: What is the behaviour when I pass a block size for the inner-most dim and `cb_inner_dim` is `False`? This looks confusing. Especially for an example.

File: devito/propagator.py [hunk 2]
```diff
@@ -244,7 +247,7 @@ def generate_loops(self, loop_body):
             time_stepping = self.get_time_stepping()
         else:
             time_stepping = []
-        loop_body = omp_for + [loop_body] if self.compiler.openmp else [loop_body]
+        loop_body = [cgen.Block(omp_for + loop_body)]
```
Comment (navjotk) [inline devito/propagator.py]: Is this change relevant to this feature?

File: devito/propagator.py [hunk 3]
```diff
@@ -281,56 +283,75 @@ def generate_space_loops(self, loop_body):
...
+                upper_limit_str = str(loop_limits[1])
+
+            loop_body = cgen.For(cgen.InlineInitializer(cgen.Value("int", dim_var), lower_limit_str),
+                                 dim_var + "<" + upper_limit_str, dim_var + "++", loop_body)
+
             if ivdep and len(self.space_dims) > 1:
                 loop_body = cgen.Block([self.compiler.pragma_ivdep] + [loop_body])
             ivdep = False
 
+        ivdep = True
```
Comment (navjotk) [inline devito/propagator.py]: Please rename this flag to something more indicative of its use: `innermost_dim` or something.

File: devito/propagator.py (hunk 2 above)
Comment (sheino) [inline devito/propagator.py]: It solves a bug. Where openmp pragma was written in form like this:
#pragma openmp for
{
for(...)

which caused compiling error

File: README.md (hunk 1 above)
Comment (sheino) [inline README.md]: it will be ignored. But well spoted the example looks confusing. Gonna update README.

File: tests/test_cache_blocking.py [hunk 4]
```diff
@@ -30,3 +30,25 @@ def test_cache_blocking_remainder(self):
...
         assert(np.equal(output_grid_block.data, output_grid_noblock.data).all())
+
+    def test_cache_blocking_full_range(self):
+        input_grid = DenseData(name="input_grid", shape=(302, 302), dtype=np.float64)
+        input_grid.data[:] = np.arange(91204, dtype=np.float64).reshape((302, 302))
+        x, t = symbols("x t")
+
+        errors = []
+        for spc_border in range(0, 9):
+            for block_size in range(1, 11):
```
Comment (mlange05) [inline tests/test_cache_blocking.py]: We should really parametrize the test itself using `pytest.mark.parametrize`, rather than looping inside the test. The advantage is that pytest will immediately tell you which combinations fail and which ones pass. We can also parametrize the "block-inner-loop" feature to keep the test setup clean.

**Comment by sheino** [comment]
Can we merge this next? As its quite an essential update cache bloking.

File: devito/propagator.py [hunk 5]
```diff
@@ -83,11 +85,14 @@ def __init__(self, name, nt, shape, spc_border=0, time_order=0,
 
         # Cache blocking and default block sizes
         self.cache_blocking = cache_blocking
-        if(isinstance(block_size, Iterable)):
-            if(len(block_size) == len(shape)):
+        self.cache_block_inner = cb_inner_dim
+        if isinstance(block_size, Iterable):
+            if len(block_size) == len(shape):
                 self.block_sizes = block_size
+            elif len(block_size) == len(shape) - 1 and not self.cache_block_inner:
```
Comment (mlange05) [inline devito/propagator.py]: Would passing a `block_size` of length `len(shape)-1` not imply that the innermost dimension is not blocked? If so, is the additional parameter `cb_inner_dim` even necessary?

File: devito/propagator.py [hunk 6]
```diff
@@ -293,56 +297,76 @@ def generate_space_loops(self, loop_body):
...
             dim_var = str(self._var_map[spc_var])
             block_var = dim_var + "b"
             loop_limits = self._space_loop_limits[spc_var]
-            loop_body = cgen.For(cgen.InlineInitializer(cgen.Value("int", dim_var),
-                                                        block_var),
-                                 dim_var + "<" + block_var+"+"+str(block_size), dim_var + "++", loop_body)
-            if ivdep and len(self.space_dims) > 1:
+
+            lower_limit_str = block_var
+            upper_limit_str = block_var+"+"+str(block_size)
```
Comment (mlange05) [inline devito/propagator.py]: Can we please be consistent with spacing around `+`, even with strings?

File: devito/propagator.py (hunk 5 above)
Comment (sheino) [inline devito/propagator.py]: Yes, it is neccessary because with current implementation we can pass single `block_size`, which would give square blocks on all dimensions. Thus the flag is required, if we do not want to block on all dimensions in that case.
I added this statement here to append dummy value because with current implementation `block_size` is zipped with `shape `and then iterated over to create blocking loops. Thus, length of them has to be the same when creating loops.

Comment (mlange05) [inline devito/propagator.py]: So, if I understand you right, we add a separate interface parameter `cb_inner_dim` just enable a shorthand notation where `block_size=n` gets broadcast to `block_size=[n, n, n]`? That doesn't sound like a particularly profitable trade-off to me. I would prefer to just pass in `block_size=[n, n]`, dummy all missing inner dimensions if `len(block_size) < len(shape)` and leave the simple broadcast for square blocks to the user application or auto-tuning script. This would reduce the complexity and special-casing in the interface, and still not require much more work from the user, IMHO.

Comment (sheino) [inline devito/propagator.py]: That is one of the reasons why I added it. 
We still need a flag when we are looping, creating `for loops` with blocking, as at that time `len(block_size) == len(shape)` even if ignoring inner most dim:

 ```
  for spc_var, block_size in reversed(zip(list(self.space_dims), self.block_sizes)):
                 dim_var = str(self._var_map[spc_var])
                 loop_limits = self._space_loop_limits[spc_var]
-                old_upper_limit = loop_limits[1]
-                new_upper_limit = old_upper_limit-old_upper_limit % block_size
-                loop_limits = (new_upper_limit, old_upper_limit)
+
+                # Does not block inner most dim if cache_block_inner flag is set
+                if not inner_most_dim or self.cache_block_inner:
+                    loop_limits = (loop_limits[1] - (loop_limits[1] - loop_limits[0]) % block_size, loop_limits[1])
```

We can't skip the iteration even though we are not cb inner most dim in this case.

Furthermore when auto tuning `block_sizes` are ignored, thus the flag is still needed as seen above to differentiate which dimensions to block/prep for auto tuning.

Comment (mlange05) [inline devito/propagator.py]: Ok, so there are a few things here I don't quite understand:
a) I am mainly concerned with the complexity of the interface to Propagator and Operator, and I see nothing in your argument that convinces me that an outward facing `cb_inner_dim` parameter is required. Why does the user need to see this exactly?
b) I also don't quite see the need for an explicit flag. Instead could you not insert `None` as the default value into `block_sizes` and check `if block_size is not None` in the loop you quote? This would also allow us more flexibility as to which dimensions to leave open, not just the inner one.
c) Just to understand this better, when auto-tuning you are injecting a separate `block_sizes` array at some point into the Propagator toolchain for each iteration in the parameter space, are you not? Otherwise, how are you re-using the existing cache blocking code?

Comment (sheino) [inline devito/propagator.py]: Hmm.. so you're saying that `block_size` parameter length could tell propagator how many dims we want to cache block/auto tune am I correct?

File: devito/propagator.py [hunk 7]
```diff
@@ -314,10 +314,10 @@ def generate_space_loops_blocking(self, loop_body):
             block_var = dim_var + "b"
             loop_limits = self._space_loop_limits[spc_var]
 
-            lower_limit_str = block_var
-            upper_limit_str = block_var+"+"+str(block_size)
-
-            if inner_most_dim and not self.cache_block_inner:
+            if block_size:
```
Comment (mlange05) [inline devito/propagator.py]: Can we please be explicit here: `if block_size is not None:`?

File: devito/propagator.py [hunk 8]
```diff
@@ -328,11 +328,9 @@ def generate_space_loops_blocking(self, loop_body):
                 loop_body = cgen.Block([self.compiler.pragma_ivdep] + [loop_body])
             inner_most_dim = False
 
-        inner_most_dim = True
         for spc_var, block_size in reversed(zip(list(self.space_dims), self.block_sizes)):
-            # if not cache blocking inner most dim skip this iteration
-            if not inner_most_dim or self.cache_block_inner:
-
+            # if block size set to None do not block this dimension
+            if block_size:
```
Comment (mlange05) [inline devito/propagator.py]: Can we please be explicit here: `if block_size is not None:`?

File: devito/propagator.py [hunk 9]
```diff
@@ -354,8 +351,8 @@ def generate_space_loops_blocking(self, loop_body):
                 dim_var = str(self._var_map[spc_var])
                 loop_limits = self._space_loop_limits[spc_var]
 
-                # Does not block inner most dim if cache_block_inner flag is set
-                if not inner_most_dim or self.cache_block_inner:
+                # Does not block this dim if block_size is None
+                if block_size:
```
Comment (mlange05) [inline devito/propagator.py]: Can we please be explicit here: `if block_size is not None:`?

File: devito/operator.py [hunk 10]
```diff
@@ -90,7 +90,7 @@ class Operator(object):
     def __init__(self, nt, shape, dtype=np.float32, stencils=[],
                  substitutions=[], spc_border=0, time_order=0,
                  forward=True, compiler=None, profile=False,
-                 cache_blocking=False, block_size=5, cb_inner_dim=False,
+                 cache_blocking=False, block_size=5,
```
Comment (mlange05) [inline devito/operator.py]: Can we now set a more realistic default value for `block_size`? 5 seems arbitrary. 

Also, is the `cache_blocking` flag still needed, or should setting `block_size = None` not be enough to disable cache blocking altogether?

Comment (sheino) [inline devito/operator.py]: Im gonna fix this in auto tuning  `block_size=None` (default). If block_size is not set. Auto tunning will try to either extract block size from report if one exists otherwise will calculate best square guess based on architecture. @pvelesko equation
Thus `cache_blocking` flag will be necessary.

File: devito/propagator.py (hunk 9 above)
Comment (sheino) [inline devito/propagator.py]: Fixed

**Comment by mlange05** [comment]
OK, this looks good to go now. Since there has been a lot of back-and-forth on this, I would like a second opinion from @navjotk though.

File: tests/test_cache_blocking.py [hunk 11]
```diff
@@ -1,32 +1,55 @@
...
+
+        output_grid_noblock = DenseData(name="output_grid", shape=shape, dtype=np.float64)
+        eq = Eq(output_grid_noblock.indexed[indexes], input_grid.indexed[indexes] + 3)
+        op_noblock = SimpleOperator(input_grid, output_grid_noblock, [eq], time_order=time_order, spc_border=spc_border)
         op_noblock.apply()
-        output_grid_block = DenseData(name="output_grid", shape=(302, 302), dtype=np.float64)
-        op_block = SimpleOperator(input_grid, output_grid_block, [eq], cache_blocking=True)
+
+        output_grid_block = DenseData(name="output_grid", shape=shape, dtype=np.float64)
+        op_block = SimpleOperator(input_grid, output_grid_block, [eq], cache_blocking=True,
```
Comment (navjotk) [inline tests/test_cache_blocking.py]: By reusing `eq` from line 47, you are also reusing `output_grid_noblock` in your expression. Is that right? Would you not want the expression here to involve `output_grid_block`?

Comment (sheino) [inline tests/test_cache_blocking.py]: Well I just parameterised the test that you have written before so I am not exactly sure.

I presumed that `output_grid_block` and `output_grid_noblock` are the same both in size and shape, so reusing in inside the `eq` is fine.

In any case test behaviour is as expected.

File: tests/test_cache_blocking.py [hunk 12]
```diff
@@ -1,32 +1,55 @@
...
+
+        size = 1
+        for element in shape:
+            size *= element
+
+        input_grid = DenseData(name="input_grid", shape=shape, dtype=np.float64)
+        input_grid.data[:] = np.arange(size, dtype=np.float64).reshape(shape)
+
+        output_grid_noblock = DenseData(name="output_grid", shape=shape, dtype=np.float64)
+        eq = Eq(output_grid_noblock.indexed[indexes], input_grid.indexed[indexes] + 3)
```
Comment (navjotk) [inline tests/test_cache_blocking.py]: Since this PR seeks to address some bug that showed up in certain orders of the wave equation, and this test is meant to test cache blocking with various spatial orders, I think it makes sense to make the test a little more realistic than just use this basic kernel. 
Since you are trying to spot bugs in higher spatial orders, this line here won't reveal them since nothing in this kernel is dependent on spatial orders. You might want to choose an equation here that changes with spatial order - one of the fwi or tti operators maybe.

Comment (sheino) [inline tests/test_cache_blocking.py]: This PR addresses the issues that appeared due to the wrong loop limit calculations and broken remainder loop logic. Spacial orders just chaganged the fact whether there was a remainder or not.

Basicaly what we care in this PR is making sure that all elements are iterated through as it would be without cache blocking. Therefore, I think simple equation is more than enough. Moreover, this runs around 250 tests which run for much longer if we use anything more complex.

File: devito/propagator.py [hunk 13]
```diff
@@ -293,56 +297,74 @@ def generate_space_loops(self, loop_body):
...
-        if remainder:
+            # if block size set to None do not block this dimension
+            if block_size is not None:
+                orig_var = str(self._var_map[spc_var])
+                dim_var = orig_var + "b"
+                loop_limits = self._space_loop_limits[spc_var]
+                old_upper_limit = loop_limits[1]                  # sets new upper limit
+                loop_limits = (loop_limits[0], loop_limits[1] - (loop_limits[1] - loop_limits[0]) % block_size)
+
+                if old_upper_limit - loop_limits[1] > 0:  # check old vs new upper limit
```
Comment (navjotk) [inline devito/propagator.py]: Could you move this condition to immediately following the for loop in line 351 below? That way you don't need to maintain a `remainder_list`. You don't need the additional `for` you introduced in line 347 either this way. You just check each dimension individually on whether it needs a remainder or not (based on this same condition). If it does, you generate the remainder. I think that would also remove both the `if` statements from lines 355 and 361 below.

**Comment by navjotk** [comment]
Two things:
- The test needs to test a real case. This test is not testing the code that was written since the kernel does not change with spatial orders.
- I think we have an excessively high number of parameters for a "simple" feature like cache blocking. Having `block_size`, `cache_blocking` and `cb_inner` as three different parameters seems too much to me. I think it could all be condensed into one single parameter. 

Apart from the one place where I commented, the code looks fine on visual inspection.

File: tests/test_cache_blocking.py (hunk 12 above)
Comment (navjotk) [inline tests/test_cache_blocking.py]: The error that was exposed at higher spatial orders had to do with loop limit calculations, correct? In that case is it really necessary to run 250 different tests for it? You could probably achieve the same with just 2 different tests.

Comment (navjotk) [inline tests/test_cache_blocking.py]: About the remainder loop calculations. You have changed the implementation of remainder loop but haven't changed the test at all. If the previous implementation of remainder was broken, how is it that it was passing the tests? If the tests were incorrect, don't you want to add a test that the previous "broken" implementation would fail but your new implementation passes?

File: tests/test_cache_blocking.py (hunk 11 above)
Comment (navjotk) [inline tests/test_cache_blocking.py]: A test that always passes is not much of a test at all :)

File: tests/test_cache_blocking.py (hunk 12 above)
Comment (sheino) [inline tests/test_cache_blocking.py]: The previous implementations of the test, was alright, and it was passing because the model was too small. This remainder loop error was only exposed when you blocked 2 or more dimensions (as you need 2 or more remainder loops), whereas the model in the previous test was blocked only in 1 dim.

**Comment by navjotk** [comment]
Already communicated in private chat but putting it here for reference. We need additional tests here that  pick out 
-  whether all indices are traversed (as was the original error that was fixed in this PR)
-  repetition of indices as pointed out

The code currently in master should fail these tests and the code in this PR should succeed. The tests as they stand currently, would pass even on current master without this PR.

**Comment by navjotk** [comment]
This branch is still missing a test as described above.

**Comment by sheino** [comment]
I replaced equation in `test_cache_blocking.py`.

Before, which only test whether all elements been iterated through:
output_grid = input_grid + 3
Now, checks also that everything has been iterated only once:
output_grid = output_grid + input_grid + 3

This should cover the case of multiple iterations through same elements

**Comment by navjotk** [comment]
Confirmed that this test fails on current master and passes on this branch, hence ensuring that its catching the bug this PR is meant to fix. Thanks.
